### DISCUSSION TITLE
How to cache using ion to prevent suspense?

### DISCUSSION THREAD

**Discussion by devnomic** [Omnistac/zedux#55] [General]
Hi,

I am trying to derive atom using ion like this.
````
const userIdAtom = atom("userId", 1);

const userProfileAtom = ion("userProfile", ({ get }) => {
  const userId = get(userIdAtom);
  const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
  const promise = fetch(url).then((result) => result.json());
  return api(promise);
});
````

While this works, it always trigger  suspense when fetching user profile.
https://codesandbox.io/s/zedux-ion-s6w4w0?file=/src/App.tsx

![1](https://github.com/Omnistac/zedux/assets/101681592/7b51b01f-995d-4a01-bc71-37f1952ea310)

If i'm using normal atom with param, this works as intended / the results are cached.
````
const userIdAtom = atom("userId", 1);

const userProfileAtom = atom("userProfile", (userId: number) => {
  const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
  const promise = fetch(url).then((result) => result.json());
  return api(promise);
});
````

https://codesandbox.io/s/zedux-ion-2-utxibv?file=/src/App.tsx

Is it possible to achieve same result using ion?
Or is it generally better to use the second approach (atom with params) when using zedux?

Thank you

**Comment by bowheart** [comment]
Hey @devnomic, yes, it sounds like the second approach is what you want in this case. This takes advantage of automatic caching - each atom instance holds a cached query result. You can then control the lifetime of these caches with [TTL](https://omnistac.github.io/zedux/docs/walkthrough/destruction#ttl). For example:

```ts
const userProfileAtom = atom("userProfile", (userId: number) => {
  const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
  const promise = fetch(url).then((result) => result.json());
  return api(promise);
}, { ttl: 60000 }); // keep unused users in memory for 1 minute
```

Ions take params too btw, after the first AtomGetters argument

```ts
const exampleIon = ion('example', ({ get }, myParam: string) => ...)
```

Without params, the atom is a singleton - only one atom instance is ever created. If you need more control over the cache, a singleton gives you that, but you'd have to implement all the caching logic manually. I've never needed this.

If you don't want to have to pass params every time, you can create a selector that does it for you:

```ts
const getUserProfile = ({ get }: AtomGetters) => get(userProfileAtom, [get(userIdAtom)])

// then in a component:
const { data } = useAtomSelector(getUserProfile)
```

I use this all the time, but (very important) note that `useAtomSelector` doesn't trigger suspense currently. It's possible to make a custom hook that does though. [Here's an example](https://codesandbox.io/s/zedux-suspending-selectors-g61k4q?file=/src/App.tsx). I've been really wanting this in Zedux proper for a while now and actually I think I'm going to spend some time working on it today.

**Reply by devnomic** [reply]
I see, thank you for the reply.

I also want to ask how about if i need to fetch the user profile periodically, kinda `refetchInterval` in react query?
Should i  do it in the when rendering component? call `useEffect` and `setInterval` and `invalidate` it? or i set it inside the atom (which i believe is the preferred way??).

Use case would be like display latest stock price

````javascript
const symbolAtom = atom("symbol", "AAPL");

// I want this atom to update/invalidate periodically
const priceAtom = atom("price", (symbol: string) => {

  injectEffect(() => {
    const intervalId = setInterval(() => {
        // How to invalidate?
        // instance.invalidate()?
    }, 5000)
    return () => clearInterval(intervalId)
  }, [])

  const url = `https://jsonplaceholder.typicode.com/users/${symbol}`;
  const promise = fetch(url).then((result) => result.json());
  return api(promise);
});
````

My imagination 😄 
````js
// I want this atom to update/invalidate periodically
const priceAtom = atom("price", (symbol: string) => {
  const url = `https://jsonplaceholder.typicode.com/users/${symbol}`;
  const promise = fetch(url).then((result) => result.json());
  return api(promise);
}, { invalidateInterval: 5000 } );
````

**Reply by devnomic** [reply]
Another question, how to make the atom typesafe for user profile?

```ts
interface UserProfile {
  name: string;
  email: string;
}

// This obviously not working
const userProfileAtom = atom<UserProfile>(
  "userProfile",
  (userId: number) => {
    const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
    const promise = fetch(url).then((result) => result.json());
    return api(promise);
  }
);
```

or perhaps like this since this works?
```ts
const userProfileAtom = atom("userProfile", (userId: number) => {
  const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
  const promise: Promise<UserProfile> = fetch(url).then((result) =>
    result.json()
  );
  return api(promise);
});
```

**Reply by bowheart** [reply]
Ooh. `invalidateInterval` is a cool idea. I'll keep that in mind for the `@zedux/query` package we're planning on making.

You basically got it. Just add [`injectSelf()`](https://omnistac.github.io/zedux/docs/api/injectors/injectSelf). Here's a custom injector example:

```ts
function injectInvalidateInterval(ms = 5000) {
  const instance = injectSelf();

  injectEffect(() => {
    const intervalId = setInterval(() => {
      instance.invalidate();
    }, ms);

    return () => clearInterval(intervalId);
  }, []);
}
```

Then use it like so:

```ts
const priceAtom = atom("price", (symbol: string) => {
  const url = `https://jsonplaceholder.typicode.com/users/${symbol}`;
  const promise = fetch(url).then((result) => result.json());

  injectInvalidateInterval(5000);

  return api(promise);
});
```

> how to make the atom typesafe

Yeah your second example is exactly how I'd do it. Zedux's types try really hard to auto-infer everything. So the easiest way to type things is to feed Zedux good types. The trade-off of that is it can be tedious to manually type some things. But I believe this'll do it if you need:

```ts
import { api, atom, PromiseState } from '@zedux/react';

const userProfileAtom = atom<
  PromiseState<UserProfile>,
  [userId: number],
  Record<string, never>
>("userProfile", (userId: number) => {
  const url = `https://jsonplaceholder.typicode.com/users/${userId}`;
  const promise = fetch(url).then(result => result.json());
  return api(promise);
});
```

Ideally you should never need to do this, but I'd be interested to know any exceptions you find. If [partial inference](https://github.com/microsoft/TypeScript/pull/54047) or [keyed generics](https://github.com/microsoft/TypeScript/issues/54254) land in TypeScript, it'll make manual typing a lot easier in Zedux.

**Reply by devnomic** [reply]
Thanks again!

**Reply by joprice** [reply]
@bowheart `useSuspendingSelector` would be a great addition! I just copied it over and made use of it. Maybe this and other ones found in the discussions could go into a utils package.

**Reply by bowheart** [reply]
A utils package is definitely something we should consider. I've made #143 to track that.
